src/creatures/players/components/player_forge_history.cpp

```diff
@@ -0,0 +1,121 @@
+/**
+ * Canary - A free and open-source MMORPG server emulator
+ * Copyright (©) 2019-2022 OpenTibiaBR <opentibiabr@outlook.com>
+ * Repository: https://github.com/opentibiabr/canary
+ * License: https://github.com/opentibiabr/canary/blob/main/LICENSE
+ * Contributors: https://github.com/opentibiabr/canary/graphs/contributors
+ * Website: https://docs.opentibiabr.org/
+ */
+
+#include "creatures/players/components/player_forge_history.hpp"
+
+#include "database/database.hpp"
+#include "creatures/players/player.hpp"
+#include "utils/tools.hpp"
+#include "game/scheduling/save_manager.hpp"
+
+PlayerForgeHistory::PlayerForgeHistory(Player &player) :
+	m_player(player) { }
+
+const std::vector<ForgeHistory> &PlayerForgeHistory::get() const {
+	return m_history;
+}
+
+void PlayerForgeHistory::add(const ForgeHistory &history) {
+	m_history.push_back(history);
+	m_modifiedHistory.push_back(history);
+}
+
+void PlayerForgeHistory::remove(int historyId) {
+	m_removedHistoryIds.push_back(historyId);
+	auto it = std::ranges::remove_if(m_history, [historyId](const ForgeHistory &h) {
+		return h.id == historyId;
+	});
+	m_history.erase(it.begin(), it.end());
+}
+
+bool PlayerForgeHistory::load() {
+	auto playerGUID = m_player.getGUID();
+	auto query = fmt::format("SELECT * FROM forge_history WHERE player_id = {}", playerGUID);
+	const DBResult_ptr &result = g_database().storeQuery(query);
+	if (!result) {
+		g_logger().debug("Failed to load forge history for player with ID: {}", playerGUID);
+		return false;
+	}
+
+	do {
+		ForgeHistory history;
+		history.id = result->getNumber<int>("id");
+		history.actionType = static_cast<ForgeAction_t>(result->getNumber<uint8_t>("action_type"));
+		history.description = result->getString("description");
+		history.createdAt = result->getNumber<uint64_t>("done_at");
+		history.success = result->getNumber<bool>("is_success");
+		m_history.push_back(history);
```

Comment:
Isn't better use emplace_back to construct directly it in place?

Reply:
You're right. Thanks!